test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx

```diff
@@ -174,8 +179,140 @@ BOOST_AUTO_TEST_CASE(test_export_coreml) {
 }
 
+BOOST_AUTO_TEST_CASE(test_load_version) {
+  turi::drawing_classifier::drawing_classifier dc;
+  TS_ASSERT_EQUALS(dc.get_version(), dc.DRAWING_CLASSIFIER_VERSION);
+}
+
+
 BOOST_AUTO_TEST_CASE(test_save_load) {
-  // TODO
+  auto remove_if_exist =
+      [](const char* fname) {
+        std::FILE* fp = nullptr;
+        if ((fp = std::fopen(fname, "r")) != NULL) {
+          if (std::fclose(fp) != 0) {
+            std::stringstream ss;
+            ss << "cannot close file: " << fname << std::endl;
+            throw(turi::error::io_error(ss.str()));
+          }
+
+          if (std::remove(fname) != 0) {
+            std::stringstream ss;
+            ss << "cannot remove file: " << fname << std::endl;
+            throw(turi::error::io_error(ss.str()));
+          }
+        }
+      };
+
+  /**
+   * This is a hack because nn_spec_ is a private member. Thus, I can't
+   * compare it directly after loading from file (load_version call).
+   *
+   * The solution is to call `load_version` first and write serialized
+   * `nn_spec_` to disk. In this way, I can indirectly verify the `nn_spec_`
+   * successively loads all params from the source model spec file.
+   */
+  auto load_save_compare = [&remove_if_exist](drawing_classifier_mock& dc,
+                              drawing_classifier_mock& dc_other) {
+    // random init; avoid segfault
+    dc.my_init_model_spec();
+    dc_other.my_init_model_spec();
+
+    constexpr auto my_file = "./test_dc_serialization.cxx.save.txt";
+    remove_if_exist(my_file);
+    std::stringstream ss;
+
+    // save impl from the first mock
+    {
+      std::fstream out_file(my_file, std::ios_base::out | std::ios_base::binary);
+      turi::oarchive oarch(out_file);
+      dc.save_impl(oarch);
+      out_file.close();
+    }
+
+    // load impl to other dc mock instance
+    {
+      std::fstream in_file(my_file, std::ios_base::in | std::ios_base::binary);
+      turi::iarchive iarch(in_file);
+      dc_other.load_version(iarch, dc_other.get_version());
+      in_file.close();
+    }
+
+    // loading params from model spec
+    // saved file will be used to compare with the model spec saved before
+    constexpr auto my_file_loaded = "./test_dc_serialization.cxx.load.txt";
+    remove_if_exist(my_file_loaded);
+
+    {
+      std::fstream out_file(my_file_loaded,
+                            std::ios_base::out | std::ios_base::binary);
+      turi::oarchive oarch(out_file);
+      dc_other.save_impl(oarch);
+      out_file.close();
+    }
+
+    // validation
+    std::ifstream file(my_file, std::ios::in | std::ios::binary | std::ios::ate);
+    if (!file.is_open()) {
+      std::stringstream ss;
+      ss << "fail to open" << my_file << std::endl;
+      throw std::ios_base::failure(ss.str());
+    }
+    std::streamsize file_size = file.tellg();
+    file.seekg(0, std::ios::beg);
+
+    std::vector<char> model_spec_src(file_size);
+    file.read(model_spec_src.data(), file_size);
+    file.close();
```

Comment:
These are a lot of low-level filesystem operations for this kind of test. @hoytak Do we have any abstractions for testing serialization code?

Alternatively, typically I would expect that we could just test that a save/load round trip preserves the model spec and state.

Reply:
These operations just tell me how large the file is. And then I load all bytes of those files and compare the contents all together to make sure we have the same results after loading.